src/audio/mux/mux.c

```diff
@@ -200,7 +204,195 @@ static struct comp_dev *mux_new(const struct comp_driver *drv,
+				     struct comp_ipc_config *config,
+				     void *spec)
+{
+	struct mux_data *ipc_process = spec;
+	struct comp_dev *dev;
+	struct comp_data *cd;
+	int ret;
 
+	comp_cl_info(&comp_mux, "mux_new()");
+
+	dev = comp_alloc(drv, sizeof(*dev));
+	if (!dev)
+		return NULL;
+	dev->ipc_config = *config;
+
+	cd = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(*cd));
+	if (!cd) {
+		rfree(dev);
+		return NULL;
+	}
+
+	comp_set_drvdata(dev, cd);
+
+	ret = memcpy_s(&cd->md, sizeof(cd->md), ipc_process, sizeof(*ipc_process));
+	assert(!ret);
+	ret = build_config(cd, dev);
+	assert(!ret);
+
+	if (ret < 0) {
+		rfree(cd);
+		rfree(dev);
+		return NULL;
+	}
+	dev->state = COMP_STATE_READY;
+	return dev;
+}
+static void set_mux_params(struct comp_dev *dev,
+				struct sof_ipc_stream_params *params)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	struct comp_buffer *sink, *source;
+	struct comp_buffer __sparse_cache *sink_c, *source_c;
+	struct list_item *source_list;
+	int ret = 0;
+	int i, j;
+	const uint32_t byte_align = 1;
+	const uint32_t frame_align_req = 1;
+
+	params->direction = dev->direction;
+	params->channels =  cd->md.base_cfg.audio_fmt.channels_count;
+	params->rate = cd->md.base_cfg.audio_fmt.sampling_frequency;
+	params->sample_container_bytes = cd->md.base_cfg.audio_fmt.depth / 8;
+	params->sample_valid_bytes = cd->md.base_cfg.audio_fmt.valid_bit_depth / 8;
+	params->buffer_fmt = cd->md.base_cfg.audio_fmt.interleaving_style;
+	params->buffer.size = cd->md.base_cfg.ibs;
+	params->no_stream_position = 1;
+
+	if (!list_is_empty(&dev->bsink_list)) {
+		sink = list_first_item(&dev->bsink_list, struct comp_buffer, source_list);
+		sink_c = buffer_acquire(sink);
+		audio_stream_init_alignment_constants(byte_align, frame_align_req,
+				__sparse_cache &sink_c->stream);
+
+		if (!sink_c->hw_params_configured) {
+			struct ipc4_audio_format out_fmt;
+
+			out_fmt = cd->md.output_format;
+			sink_c->stream.channels = out_fmt.channels_count;
+			sink_c->stream.rate = out_fmt.sampling_frequency;
+			audio_stream_fmt_conversion(out_fmt.depth,
+							out_fmt.valid_bit_depth,
+							&sink_c->stream.frame_fmt,
+							&sink_c->stream.valid_sample_fmt,
+							out_fmt.s_type);
+
+			sink_c->buffer_fmt = out_fmt.interleaving_style;
+			params->frame_fmt = sink_c->stream.frame_fmt;
+
+			for (i = 0; i < SOF_IPC_MAX_CHANNELS; i++)
+				sink_c->chmap[i] = (out_fmt.ch_map >> i * 4) & 0xf;
+
+			sink_c->hw_params_configured = true;
+		}
+		buffer_release(sink_c);
+	}
+
+
+	if (!list_is_empty(&dev->bsource_list)) {
+		list_for_item(source_list, &dev->bsource_list)
+		{
+			source = container_of(source_list, struct comp_buffer, sink_list);
+			source_c = buffer_acquire(source);
+			audio_stream_init_alignment_constants(byte_align, frame_align_req,
+					__sparse_cache &source_c->stream);
+			j = source_c->id;
+			cd->config.streams[j].pipeline_id = source_c->pipeline_id;
+			if (j == 0) {
```

Comment:
Comment would help, why does "j == 0" get special treatment?

Reply:
struct mux_data {
	struct ipc4_base_module_cfg base_cfg;
	struct ipc4_audio_format reference_format;
	struct ipc4_audio_format output_format;
};
In the mux module we have two inputs and one output, for the first input we assign parameters from base cfg, for the second from reference_format, and for sink output_format